src/main/java/de/tum/in/www1/artemis/service/StaticCodeAnalysisService.java

```diff
@@ -98,6 +104,26 @@ public void createDefaultCategories(ProgrammingExercise programmingExercise) {
             originalCategory.setState(matchingCategory.getState());
         }
         staticCodeAnalysisCategoryRepository.saveAll(originalCategories);
+
+        // At least one category was updated. We use this flag to inform the instructor about outdated student results.
+        programmingSubmissionService.setTestCasesChangedAndTriggerTestCaseUpdate(exerciseId);
+
         return originalCategories;
     }
+
+    /**
+     * Links the categories of an exercise with the default category mappings.
+     * @param programmingExercise The programming exercise
+     * @return A list of pairs of categories and their mappings.
+     */
+    public List<ImmutablePair<StaticCodeAnalysisCategory, List<StaticCodeAnalysisDefaultCategory.CategoryMapping>>> getCategoriesWithMappingForExercise(
+            ProgrammingExercise programmingExercise) {
+        var categories = findByExerciseId(programmingExercise.getId());
+        var defaultCategories = staticCodeAnalysisDefaultConfigurations.get(programmingExercise.getProgrammingLanguage());
+
+        return categories.stream()
```

Comment:
this is unreadable and not understandable. Please do not abuse the stream API and simplify this code so that it is easier to understand!

Reply:
I improved the code readability here.